### PR TITLE
Avatarabo/firemario

### LINKED ISSUE #15: Adding a Fire state to Mario
Implement all the interaction and turtle shell features of Koopa. Here is some information based on [Mario Wiki](https://www.mariowiki.com/Fire_Mario#Super_Mario_Bros._/_Super_Mario_Bros.:_The_Lost_Levels).

1. ~~Hats and overalls change to a light cream color, while the shirt turns red.~~
2. If Mario or Luigi are in this form when the timer runs out, their losing sprites will use the fire palette.
3. ~~Received this state after obtaining a Fire Flower.~~
4. ~~Can throw fireballs (up to 2 on-screen at a time).~~.
5. The fireballs bounce along the ground and can defeat almost all enemies, including otherwise unbeatable ones such as Spinies
6. Using five fireballs to defeat Bowser and his copies is the only method that yields points (5000 points).
7. Turn back to small form (normal form) if takes damage.
8. If a Fire Flower is spawned but the player is reverted to their Small form before collecting it, it will merely change them to their Super form.

Update (Feb. 18, 2021): This issue is keep opened due to in completed tasks (strike through). The implementation of No.4 contains potential bugs as mentioned (#25). No.5 is partially implemented (#25). No. 8 is not tested (#25).

### PR DESCRIPTION
**Pull request by AvatarAbo**
# Pull Request Description

Please fill out the form below. For the checkbox, use [x] to have a checkmark. For example:

- [x] With a checkmark.
- [ ] Without a checkmark.

This should not affect the way whether your pull request is accepted or not. Think of it like a survey and documentation, so the reviewers and future developers can follow your PR easier.

1. Have you included *.gitignore* file? Or manually remove config and environment files (E.g: *node_modules* folder, Eclipse IDE files) from this submission? The answer should be **Yes** in this case.

   - [x] Yes
   - [ ] No

2. Did you fully complete the assigned task(s)?

   - [ ] Fully completed
   - [x] Partially completed

3. Did you introduce a new non-required feature in this PR? This feature is not an assigned task, but it is the one that you feel will make the game better. For example, reproduce a bug used by speed runner in an original version.

   - [ ] Yes
   - [x] No

4. Are you aware of any bug caused by your code that have not been fixed in this PR?

   - [ ] Yes
   - [x] No

5. Are you aware of any **potential bug**? You don't have to fix this, it is just a way to document. You may not face the bug during the development process, but theoretically it can happens in edge cases. For example, memory leak or a player falls out of map.

   - [x] Yes
   - [ ] No

## Unimplemented task(s)

- *Only answer when you chose **Partially completed** in question 2, provide a brief summary of what you have not done in here.*

Unimplemented Tasks:
- if Mario or Luigi is in this form when the timer runs out, their losing sprites will use the fire palette
- using five fireballs to defeat Bowser and his copies is the only method that yields points (5000 points)
- turn back to small form (normal form) if Mario takes damage
- fireballs only kill goombas and koopas, but not other enemies

## Related issue(s)

- adding a fire state to Mario

## Added

- Hats and overalls change to a light cream color, while the shirt turns red.
- Received this state after obtaining a fire flower
- can throw fireballs (up to 2 on-screen at a time).
- the fireballs bounce along the ground and can defeat goombas and koopas
- If a fire flower is spawned but the player is reverted to their small form before collecting it, it will merely change them to their Super form (it will show up as a growth mushroom, not a fire flower)

## Changed and Enhanced

none

## Fixed

none

## Bugs

1. Not-fixed bugs (only answer when you chose **Yes** in question 4):
   - not applicable
2. Potential bugs (only answer when you chose **Yes** in question 5):
   - physics of fireball speed and bounce is estimated
   - fire flower animations speed is estimated
   - fire balls will be removed from world after this.y is > 768 (the current canvas height) so basically once the fireball goes off the bottom of the screen it is removed from the world so that Mario can throw more fireballs since his limit is 2 fireballs

## Contributors and code author(s)

- Albert Lin (avatarabo)

## Final check

Have you fully filled out this document, and add a label **PR: NEED REVIEW** to this pull request? Answer *Yes* (as it should be) means you are good to go! Thanks for the contribution.

- [x] Yes
- [ ] No

### CODE REVIEW

File: items.js [hunk 1]
```diff
@@ -146,4 +146,143 @@ class Mushroom {
...
+        this.animation.drawFrame(this.game.clockTick, ctx, this.x - this.game.camera.x, this.y, PARAMS.SCALE);
+        if (this.emerging) this.brick.draw(ctx);
+        if (PARAMS.DEBUG) {
+            ctx.strokeStyle = 'Red';
+            ctx.strokeRect(this.BB.x - this.game.camera.x, this.BB.y, this.BB.width, this.BB.height);
+        }
+    };
+};
+
+class Fireball {
```
Comment (BenjaminDeJager) [inline items.js:191]: this class name does NOT conflict with the "fire" class that seoungdeok added for their "firebar" implementation (fortunetely), a enemy that makes fireballs float/rotate around a brick, figured I'd check that their not the same and alert the two of you to NOT change the name for that reason.

**Review by BenjaminDeJager — APPROVED** [review]
this pr is not fully filled out, I will ask hung to decide if he wants this pr filled out but the code itself looks fine.

if it were just me I would ask that you fill out the pr just for consistency, but since I see that I'm the first reviewer I will let whoever comes after me decide if it should be required since they can either approve and merge it or request changes to your pr-form.

**Comment by hunghvu** [comment]
@AvatarAbo I would advise you to fully fill out the form fully so we can make the PRs consistent. I will review this PR tomorrow.

**Review by hunghvu — APPROVED** [review]
Overall, looks good to me, but I cannot test No. 8 of #15 because I cannot find a way to revert Mario back to his original form. I will approve this PR. However,
1. The fireball should disappear when it's out of canvas, not just (y > canvas.height). You already noticed this in potential bugs too.
2. Take a look at this picture. When throwing a fire ball, the fireball does not collide with the cliff side, only its top. I believe it is not the case with an original game. Any thought? @BenjaminDeJager 
image

**Pull request merged**